File: kokokeyboard/src/main/java/com/rowland/kokokeyboard/KeyInputMethodService.java
package com.rowland.kokokeyboard;

import android.inputmethodservice.InputMethodService;
import android.inputmethodservice.Keyboard;
import android.inputmethodservice.KeyboardView;
import android.text.TextUtils;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;

//ToDo: allow  integration of custom keyboard as a systme wide keyboard
public class KeyInputMethodService extends InputMethodService implements KeyboardView.OnKeyboardActionListener {

    private Keyboard mKeyboard;
    private Keyboard mKeyboardWithNum;
    private Keyboard mActiveKeyboard;

    private StringBuilder mComposing = new StringBuilder();
    private KeyboardView mKeyboardInputView;

    @Override
    public void onCreate() {
        super.onCreate();

    }

    @Override
    public void onInitializeInterface() {
        mKeyboard = new Keyboard(this, R.xml.keyboard_pad);
        mKeyboardWithNum = new Keyboard(this, R.xml.keyboard_with_num);
    }

    @Override
    public void onStartInput(EditorInfo attribute, boolean restarting) {
        super.onStartInput(attribute, restarting);

        switch (attribute.inputType & EditorInfo.TYPE_MASK_CLASS) {
            case EditorInfo.TYPE_CLASS_NUMBER:
                mActiveKeyboard = mKeyboardWithNum;
                break;
            case EditorInfo.TYPE_TEXT_VARIATION_PERSON_NAME:
                mActiveKeyboard = mKeyboard;
                break;
            case EditorInfo.TYPE_CLASS_TEXT:
                mActiveKeyboard = mKeyboardWithNum;
                break;

            default:
                mActiveKeyboard = mKeyboardWithNum;

        }
    }

    @Override
    public void onStartInputView(EditorInfo attribute, boolean restarting) {
        super.onStartInputView(attribute, restarting);

        mKeyboardInputView.closing();
        mKeyboardInputView.setKeyboard(mActiveKeyboard);
        mComposing.setLength(0);
    }


    @Override
    public void onFinishInput() {
        super.onFinishInput();

        if (mKeyboardInputView != null) {
            mKeyboardInputView.closing();
        }
    }


    @Override
    public View onCreateInputView() {
        mKeyboardInputView = (KeyboardView) getLayoutInflater().inflate(R.layout.qwerty_keypad, null);
        mKeyboardInputView.setOnKeyboardActionListener(this);
        mKeyboardInputView.setKeyboard(mActiveKeyboard);
        mKeyboardInputView.setPreviewEnabled(false);
        return mKeyboardInputView;
    }

    @Override
    public void onPress(int primaryCode) {

    }

    @Override
    public void onRelease(int primaryCode) {

    }

    @Override
    public void onKey(int primaryCode, int[] keyCodes) {

        switch (primaryCode) {
            case Keyboard.KEYCODE_DELETE:
                CharSequence selectedText = getCurrentInputConnection().getSelectedText(0);
                if (TextUtils.isEmpty(selectedText)) {
                    getCurrentInputConnection().deleteSurroundingText(1, 0);
                } else {
                    getCurrentInputConnection().commitText("", 1);
                }
                break;

            case Keyboard.KEYCODE_DONE:
                getCurrentInputConnection().sendKeyEvent(new KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_ENTER));
            case Keyboard.KEYCODE_MODE_CHANGE:
                if (mKeyboardInputView != null) {
                    Keyboard current = mKeyboardInputView.getKeyboard();
                    if (current == mKeyboard) {
                        mKeyboardInputView.setKeyboard(mKeyboardWithNum);
                    } else {
                        mKeyboardInputView.setKeyboard(mKeyboard);
                    }
                }
                break;
            default:
                getCurrentInputConnection().commitText(String.valueOf((char) primaryCode), 1);
        }
    }

    @Override
    public void onText(CharSequence text) {

    }

    @Override
    public void swipeLeft() {

    }

    @Override
    public void swipeRight() {

    }

    @Override
    public void swipeDown() {

    }

    @Override
    public void swipeUp() {

    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        switch (keyCode) {
            case KeyEvent.KEYCODE_BACK:
                if (event.getRepeatCount() == 0 && mKeyboardInputView != null) {
                    if (mKeyboardInputView.handleBack()) {
                        return true;
                    }
                }
                break;

            case KeyEvent.KEYCODE_DEL:
                if (mComposing.length() > 0) {
                    onKey(Keyboard.KEYCODE_DELETE, null);
                    return true;
                }
                break;
            case KeyEvent.KEYCODE_SPACE:
                return false;
            default:
        }
        return super.onKeyDown(keyCode, event);
    }
}

File: app/src/main/java/app/keyboardly/dev/keyboard/KeyInputMethodService.kt
package app.keyboardly.dev.keyboard

import android.annotation.SuppressLint
import android.inputmethodservice.InputMethodService
import android.inputmethodservice.Keyboard
import android.inputmethodservice.KeyboardView
import android.text.TextUtils
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.EditorInfo
import app.keyboardly.dev.R

//ToDo: allow  integration of custom keyboard as a system wide keyboard
class KeyInputMethodService : InputMethodService(), KeyboardView.OnKeyboardActionListener {
    private var mKeyboard: Keyboard? = null
    private var mKeyboardWithNum: Keyboard? = null
    private var mActiveKeyboard: Keyboard? = null
    private val mComposing = StringBuilder()
    private var mKeyboardInputView: KeyboardView? = null

    override fun onInitializeInterface() {
        mKeyboard = Keyboard(this, R.xml.keyboard_pad)
        mKeyboardWithNum = Keyboard(this, R.xml.keyboard_with_num)
        instance = this
    }

    override fun onStartInput(attribute: EditorInfo, restarting: Boolean) {
        super.onStartInput(attribute, restarting)
        mActiveKeyboard =
            when (attribute.inputType and EditorInfo.TYPE_MASK_CLASS) {
                EditorInfo.TYPE_CLASS_NUMBER -> mKeyboardWithNum
                EditorInfo.TYPE_TEXT_VARIATION_PERSON_NAME -> mKeyboard
                EditorInfo.TYPE_CLASS_TEXT -> mKeyboardWithNum
                else -> mKeyboardWithNum
            }
    }

    override fun onStartInputView(attribute: EditorInfo, restarting: Boolean) {
        super.onStartInputView(attribute, restarting)
        mKeyboardInputView!!.closing()
        mKeyboardInputView!!.keyboard = mActiveKeyboard
        mComposing.setLength(0)
    }

    override fun onFinishInput() {
        super.onFinishInput()
        if (mKeyboardInputView != null) {
            mKeyboardInputView!!.closing()
        }
    }

    override fun onCreateInputView(): View {
        mKeyboardInputView = layoutInflater.inflate(R.layout.default_keyboard, null) as KeyboardView
        mKeyboardInputView!!.setOnKeyboardActionListener(this)
        mKeyboardInputView!!.keyboard = mActiveKeyboard
        mKeyboardInputView!!.isPreviewEnabled = false
        return mKeyboardInputView!!
    }

    override fun onPress(primaryCode: Int) {}
    override fun onRelease(primaryCode: Int) {}
    override fun onKey(primaryCode: Int, keyCodes: IntArray?) {
        when (primaryCode) {
            Keyboard.KEYCODE_DELETE -> {
                val selectedText = currentInputConnection.getSelectedText(0)
                if (TextUtils.isEmpty(selectedText)) {
                    currentInputConnection.deleteSurroundingText(1, 0)
                } else {
                    currentInputConnection.commitText("", 1)
                }
            }
            Keyboard.KEYCODE_DONE -> {
                currentInputConnection.sendKeyEvent(
                    KeyEvent(
                        KeyEvent.ACTION_DOWN,
                        KeyEvent.KEYCODE_ENTER
                    )
                )
                if (mKeyboardInputView != null) {
                    val current = mKeyboardInputView!!.keyboard
                    if (current === mKeyboard) {
                        mKeyboardInputView!!.keyboard = mKeyboardWithNum
                    } else {
                        mKeyboardInputView!!.keyboard = mKeyboard
                    }
                }
            }
            Keyboard.KEYCODE_MODE_CHANGE -> if (mKeyboardInputView != null) {
                val current = mKeyboardInputView!!.keyboard
                if (current === mKeyboard) {
                    mKeyboardInputView!!.keyboard = mKeyboardWithNum
                } else {
                    mKeyboardInputView!!.keyboard = mKeyboard
                }
            }
            else -> currentInputConnection.commitText(primaryCode.toChar().toString(), 1)
        }
    }

    override fun onText(text: CharSequence) {}
    override fun swipeLeft() {}
    override fun swipeRight() {}
    override fun swipeDown() {}
    override fun swipeUp() {}
    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        when (keyCode) {
            KeyEvent.KEYCODE_BACK -> if (event.repeatCount == 0 && mKeyboardInputView != null) {
                if (mKeyboardInputView!!.handleBack()) {
                    return true
                }
            }
            KeyEvent.KEYCODE_DEL -> if (mComposing.isNotEmpty()) {
                onKey(Keyboard.KEYCODE_DELETE, null)
                return true
            }
            KeyEvent.KEYCODE_SPACE -> return false
            else -> {}
        }
        return super.onKeyDown(keyCode, event)
    }

    companion object {
        @SuppressLint("StaticFieldLeak")
        private var instance: KeyInputMethodService? = null


    }
}